service/src/parentchain_block_syncer.rs

```diff
@@ -0,0 +1,109 @@
+use itp_enclave_api::{enclave_base::EnclaveBase, sidechain::Sidechain};
+use itp_types::SignedBlock;
+use log::{error, trace};
+use my_node_runtime::{BlockNumber, Header};
+use sp_core::sr25519;
+use substrate_api_client::{rpc::WsRpcClient, Api};
+
+const BLOCK_SYNC_BATCH_SIZE: u32 = 1000;
+
+pub(crate) struct ParentchainBlockSyncer {
+	api: Api<sr25519::Pair, WsRpcClient>,
+}
+
+impl ParentchainBlockSyncer {
+	pub fn new(api: Api<sr25519::Pair, WsRpcClient>) -> Self {
+		ParentchainBlockSyncer { api }
+	}
+
+	/// Fetches the amount of blocks to sync from the parentchain and feeds them to the enclave.
+	/// Returns the latest synced block Header.
+	///
+	pub fn sync_parentchain<E: EnclaveBase + Sidechain>(
+		&self,
+		enclave_api: &E,
+		last_synced_header: Header,
+	) -> Header {
+		let tee_accountid = enclave_account(enclave_api);
+		trace!("Getting current head");
+		let curr_head: SignedBlock = self.api.last_finalized_block().unwrap().unwrap();
+		let head_block_number = curr_head.block.header.number;
+
+		let mut synced_header_until = last_synced_header;
+		loop {
+			let block_chunk_to_sync =
+				self.get_block_chunk_to_sync(&synced_header_until.number, &curr_head);
+			println!("[+] Found {} block(s) to sync", block_chunk_to_sync.len());
+			if block_chunk_to_sync.is_empty() {
+				return synced_header_until
+			}
+
+			let tee_nonce = self.api.get_nonce_of(&tee_accountid).unwrap();
+			if let Err(e) = enclave_api.sync_parentchain(block_chunk_to_sync.as_slice(), tee_nonce)
+			{
+				error!("{:?}", e);
+				// enclave might not have synced
+				return synced_header_until
+			};
+			synced_header_until = block_chunk_to_sync
+				.last()
+				.map(|b| b.block.header.clone())
+				.expect("Chunk can't be empty; qed");
+			println!(
+				"Synced {} out of {} finalized parentchain blocks",
+				synced_header_until.number, head_block_number,
+			)
+		}
+	}
+
+	/// gets a list of blocks that need to be synced, ordered from oldest to most recent header
+	/// starting from the last synced head until the current head but at most BLOCK_SYNC_BATCH_SIZE.
+	fn get_block_chunk_to_sync(
+		&self,
+		last_synced_head_number: &BlockNumber,
+		curr_block: &SignedBlock,
+	) -> Vec<SignedBlock> {
+		let head = curr_block.clone();
+		let no_blocks_to_sync = curr_block.block.header.number - last_synced_head_number;
+		if no_blocks_to_sync > 1 {
+			println!("light client is synced until block: {:?}", last_synced_head_number);
+			println!("Last finalized block number: {:?}\n", head.block.header.number);
+		}
+		if no_blocks_to_sync > BLOCK_SYNC_BATCH_SIZE {
+			println!(
+				"Remaining blocks to fetch until last synced header: {:?}",
+				head.block.header.number - last_synced_head_number
+			);
+			return self
+				.api
+				.get_blocks(
+					last_synced_head_number + 1,
+					last_synced_head_number + BLOCK_SYNC_BATCH_SIZE,
+				)
+				.unwrap()
+		}
+		self.api
+			.get_blocks(last_synced_head_number + 1, curr_block.block.header.number)
+			.unwrap()
+	}
```

Comment:
Here you could safe some lines with:

```
self.api
   .get_blocks(
       last_synced_header_number +1 ,
       min(last_synced_header_number + Batch_size, curr_block_number)
).unwrap()

```

And lift it out of the if statement.

Reply:
Hmm, after all these improvements, I think we can actually get rid of the method `get_block_chunt_to_sync`, and only use the snippet of the above comment in the calling loop in instead. What do you think?

Reply:
yeah I thought this too, but smashing everything in one method becomes quite lengthy.